### PR TITLE
TASK: Cleanup usage of `NodeTypeConstraintsWithSubNodeTypes`

### PR DESCRIPTION
**Pull request by mhsdesign**
The `NodeTypeConstraintsWithSubNodeTypes` is internal and should not be used outside of the corresponding sub package. We also use the `ConstraintCheck` helper for childnodes instead of expanding the types.

see also https://github.com/neos/neos-development-collection/pull/4551#discussion_r1368849813

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Review by kitsunet — APPROVED** [review]

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php [hunk 1]
```diff
@@ -216,18 +217,15 @@ protected function requireNodeTypeToAllowNodesOfTypeInReference(
...
             throw ReferenceCannotBeSet::becauseTheNodeTypeDoesNotDeclareIt($referenceName, $nodeTypeName);
         }
-        if (isset($propertyDeclaration['constraints']['nodeTypes'])) {
-            $nodeTypeConstraints = NodeTypeConstraintsWithSubNodeTypes::createFromNodeTypeDeclaration(
-                $propertyDeclaration['constraints']['nodeTypes'],
-                $this->getNodeTypeManager()
+
+        $constraints = $propertyDeclaration['constraints']['nodeTypes'] ?? [];
+
+        if (!ConstraintCheck::create($constraints)->isNodeTypeAllowed($nodeType)) {
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:223]: What is the behavior of an empty/missing `constraints` config?

**Review by bwaidelich — COMMENTED** [review]
It's a bit scary that this change does not make tests fail..

IMO we should at least add a test case to cover the behavior of missing/empty `constraints`

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:223]:
```suggestion
        if ($this->constraints === []) {
            return true;
        }
```

yes works .. and this behavior is also covered by childNodes tests

**Review by bwaidelich — APPROVED** [review]
